src/extensions/default/JavaScriptCodeHints/node/ExtractFileContent.js

```diff
@@ -0,0 +1,94 @@
+/*
+ * Copyright (c) 2017 - present Adobe Systems Incorporated. All rights reserved.
+ *  
+ * Permission is hereby granted, free of charge, to any person obtaining a
+ * copy of this software and associated documentation files (the "Software"), 
+ * to deal in the Software without restriction, including without limitation 
+ * the rights to use, copy, modify, merge, publish, distribute, sublicense, 
+ * and/or sell copies of the Software, and to permit persons to whom the 
+ * Software is furnished to do so, subject to the following conditions:
+ *  
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *  
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, 
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER 
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING 
+ * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER 
+ * DEALINGS IN THE SOFTWARE.
+ * 
+ */
+
+/*eslint-env node */
+/*jslint node: true */
+
+"use strict";
+        
+var fs = require("fs"),
+    _dirtyFilesCache = {};
+
+/**
+ * Clears the cache for dirty file paths
+ */
+function clearDirtyFilesCache() {
+    _dirtyFilesCache = {};
+}
+
+/**
+ * Updates the files cache with fullpath when dirty flag changes for a document
+ * If the doc is being marked as dirty then an entry is created in the cache
+ * If the doc is being marked as clean then the corresponsing entry gets cleared from cache
+ *
+ * @param {String} name - fullpath of the document
+ * @param {boolean} action - whether the document is dirty
+ */
+function updateDirtyFilesCache(name, action) {
+    if (action) {
+        _dirtyFilesCache[name] = true;
+    } else {
+        if (_dirtyFilesCache[name]) {
+            delete _dirtyFilesCache[name];
+        }
+    }
+}
+
+/**
+ * Extract content locally from the file system used fs.readFile()
+ *
+ * @param {String} fileName - fullpath of the document
+ * @param {Function} callback - callback handle to post the content back
+ */
+function _readFile(fileName, callback) {
+    fs.readFile(fileName, "utf8", function (err, data) {
```

Comment:
Precisely I am in a catch 22 situation 😢 . Encoding scheme is only available after detection at she'll level, in order to use that, I have to read all the files by hopping on the Brackets main thread, which negates any gain in terms of performance compared to using nodes io. I was relying on Brackets only to fetch dirty files content, but If I start relying on Brackets we will end up with performance issues in case of flat project structure. Prime pump will clog the main thread which in turn will make the entire editor non responsive.